### PR TITLE
Multi-precision support + automatic installation of numpy and cython

### PR DESCRIPTION
**Pull request by aragilar**
This is a continuation of #59 (and completely replaces the need for #54). I've switched to using a single setup.py, as the numpy recursive setup doesn't work with the cython code. There's a bit of logic so that numpy and cython will be installed if needed, even when running older versions of pip (numpy is the flaky part, due to how f2py works, cython should just work). The mutli-precision code logic should just work, I'll add some documentation about how to use it. Includes #72 for the same reason as #73 (I probably should have included #73 also, so all the tests would pass, and there'd be a single PR, let me know if you want me to do that).

### CODE REVIEW

File: apidocs/sundials.rst [hunk 1]
```diff
@@ -0,0 +1,28 @@
+Sundials Solver Options
+#######################
+
+Selecting Precision
+-------------------
+Sundials can be built with different precisions, and scikits.odes supports this.
```
Comment (bmcage) [inline apidocs/sundials.rst]: I would suggest to add here what they are: single, double or extended, with double the default.

File: ci_support/install_sundials.sh [hunk 2]
```diff
@@ -4,13 +4,16 @@ set -ex
...
+PRECISION="${SUNDIALS_PRECISION:-'double'}"
 
 wget "$SUNDIALS_URL"
 
 tar -xzvf "$SUNDIALS_FILE"
 
 mkdir sundials_build
 
+
+# ARKODE and examples are disabled as they does not support multiple precisions (will be fixed in next sundials release)
```
Comment (bmcage) [inline ci_support/install_sundials.sh]: instead of next, write fixed in sundias 3.0.0 release

File: ci_support/ensure_sundials_installed.sh [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 SUNDIALS_DEFAULT_VERSION='2.7.0'
 
-export SUNDIALS_DIR=$HOME/sundials/"${SUNDIALS_VERSION:-$SUNDIALS_DEFAULT_VERSION}"
+export SUNDIALS_DIR=$HOME/sundials/"${SUNDIALS_VERSION:-$SUNDIALS_DEFAULT_VERSION}"/"${SUNDIALS_PRECISION:-'double'}"
```
Comment (bmcage) [inline ci_support/ensure_sundials_installed.sh]: If I test this on my PC, I see eg 

    Installing: /home/benny/sundials/3.1.0/'double'/lib/libsundials_kinsol.a

Would without the apostrophe 'double' not be better?

Comment (aragilar) [reply ci_support/ensure_sundials_installed.sh]: Yeah, apparently I double quoted, fixed.

**Comment by aragilar** [comment]
I've rebased this on top of master (so #72 and #73 aren't included any more, it's just the actual changes), and fixed some typos and expanded on the types, let me know if there's anything else.

**Comment by bmcage** [comment]
@aragilar I merged but believe there is still an error. 
In https://github.com/bmcage/odes/blob/master/scikits/odes/sundials/c_sundials.pxd#L2 we hardcode for the interface realtype as float. But this should depend on the precision...

No errors in the tests because we don't actually test where it makes a difference? 
How to fix? That should be a variable line written by setup.py depending on DTYPE?

**Comment by bmcage** [comment]
Searching in sundials 3.1.1 code, I see 3 places where macro is used:

    ~/sundials/3.1.1$ grep -r "defined(SUNDIALS_SINGLE_PRECISION)"
    double/int32_t/include/sunlinsol/sunlinsol_lapackband.h:#elif defined(SUNDIALS_SINGLE_PRECISION)
    double/int32_t/include/sunlinsol/sunlinsol_lapackdense.h:#elif defined(SUNDIALS_SINGLE_PRECISION)
   double/int32_t/include/sundials/sundials_types.h:#if defined(SUNDIALS_SINGLE_PRECISION)

In 2.7.0 sunlinsol_ files will be other files but normally also present. Don't have 2.7.0 unpacked at the moment.

**Comment by aragilar** [comment]
The use of float is apparently fine when we're importing from header files, as cython apparently deals with that in code generation (according to http://docs.cython.org/en/latest/src/userguide/external_C_code.html#referencing-c-header-files), and I've been using the multi-precision code so I'm sure it works (though it can't hurt to send an email to the cython devs to check in our case).

**Comment by bmcage** [comment]
Great! I'll probably add something in comment to c_sundials.pxd like "cython doesn’t need to match the type exactly, just use something of the right general kind, correct type present in header file will be used, see cython manual"
In order to avoid having this discussion again every xx years

**Pull request merged**